extendr-api/src/iter.rs

```diff
@@ -55,22 +55,16 @@ impl StrIter {
     }
 }
 
-// Get a string reference from a CHARSXP
-fn str_from_strsxp<'a>(sexp: SEXP, index: isize) -> &'a str {
+// Get a string reference from a `CHARSXP`
+pub(crate) fn str_from_strsxp<'a>(sexp: SEXP, index: isize) -> &'a str {
     single_threaded(|| unsafe {
-        if index < 0 || index >= Rf_xlength(sexp) {
+        let charsxp = STRING_ELT(sexp, index);
```

Comment:
Why did you remove bounds check? This is C-level vulnerability

Reply:
Good question. First, `STRING_ELT` does detection on its own. Secondly, we control the state (and especially initial state) of the iterators, and we should really just define how they are instantiated correctly, instead of having checks everywhere.

I just checked, because I forgot this.. `STRING_ELT` return a `R_NilValue` if you access something outside the bounds. I think we should just handle that case and not have range checks everywhere, especially since R does it already.

Reply:
I've added and changed things. I think this works now. But I have a problem with the code organisation... maybe you can suggest a good order (wrapping into functions)?

Reply:
Do you have a source code/document reference for that? Because if you actually _checked_ it, you could have just read zeroed memory, which was interpreted as `NULL` or something like that

Reply:
I looked at the source code. I have to do that, because nothing is documented. And it checked by these intricate/altrep secure ways.. But the think is, you get `NilValue` back. Which is distinctly different than C's `NULL` pointer. I don't get C `NULL`. I get `R_NilValue` aka R `NULL`.